### ISSUE TITLE
Add documentation on interpreting health check failures?

### ISSUE THREAD

**Issue by Daenyth** [HypothesisWorks/hypothesis-python#389]
Through using hypothesis at work, I've seen a lot of confusion come from health check failures. It's very black-box to people (at least here) and I usually need to walk through it with them to help them know how to fix it.

I'm planning on making some internal documentation for this as they come up but there's no reason it really has to be internal - how can I best submit things? It looks like you're not currently using the github wiki; is this something you'd want in a page for readthedocs?

**Comment by alexwlchan** [comment]
There's already a health check page in the documentation, albeit quite a short one:

https://github.com/HypothesisWorks/hypothesis-python/blob/master/docs/healthchecks.rst

So if you were willing to share, I'm sure a pull request to add more detail to that page would be welcome.

**Comment by Daenyth** [comment]
Part of the issue is that pytest's default traceback printout is very awkward with the health checks.

Would it be possible to detect when being run by pytest and use `pytest.fail` instead of just raising the exception? This is what it looks like with pytest's default traceback handling:

```
arguments = ()
kwargs = {'self': <cmp.integration_test.reporting.test_attribution_rollup.TestAttributionRollup testMethod=test_opp_win_attribution_rollup>}
random = <random.Random object at 0x5542e40>
sd = <module 'hypothesis.strategies' from '/home/ubuntu/workspace/CMP-Python-Tox@2/...x/py27-integration/local/lib/python2.7/site-packages/hypothesis/strategies.py'>
selfy = <cmp.integration_test.reporting.test_attribution_rollup.TestAttributionRollup testMethod=test_opp_win_attribution_rollup>
given_specifier = tuples(just(()), fixed_dictionaries({'data': opp_win_events()}).map(lambda args: dict(args, **kwargs)))
perform_health_check = True
TestData = <class 'hypothesis.internal.conjecture.data.TestData'>

    @impersonate(test)
    @copy_argspec(
        test.__name__, argspec
    )
    def wrapped_test(*arguments, **kwargs):
        settings = wrapped_test._hypothesis_internal_use_settings
        if wrapped_test._hypothesis_internal_use_seed is not None:
            random = Random(
                wrapped_test._hypothesis_internal_use_seed)
        elif settings.derandomize:
            random = Random(function_digest(test))
        else:
            random = new_random()
    
        import hypothesis.strategies as sd
    
        selfy = None
        arguments, kwargs = convert_positional_arguments(
            wrapped_test, arguments, kwargs)
    
        # If the test function is a method of some kind, the bound object
        # will be the first named argument if there are any, otherwise the
        # first vararg (if any).
        if argspec.args:
            selfy = kwargs.get(argspec.args[0])
        elif arguments:
            selfy = arguments[0]
        test_runner = new_style_executor(selfy)
    
        for example in reversed(getattr(
            wrapped_test, 'hypothesis_explicit_examples', ()
        )):
            if example.args:
                if len(example.args) > len(original_argspec.args):
                    raise InvalidArgument(
                        'example has too many arguments for test. '
                        'Expected at most %d but got %d' % (
                            len(original_argspec.args), len(example.args)))
                example_kwargs = dict(zip(
                    original_argspec.args[-len(example.args):],
                    example.args
                ))
            else:
                example_kwargs = example.kwargs
            if Phase.explicit not in settings.phases:
                continue
            example_kwargs.update(kwargs)
            # Note: Test may mutate arguments and we can't rerun explicit
            # examples, so we have to calculate the failure message at this
            # point rather than than later.
            message_on_failure = 'Falsifying example: %s(%s)' % (
                test.__name__, arg_string(test, arguments, example_kwargs)
            )
            try:
                with BuildContext() as b:
                    test_runner(
                        None,
                        lambda data: test(*arguments, **example_kwargs)
                    )
            except BaseException:
                report(message_on_failure)
                for n in b.notes:
                    report(n)
                raise
        if settings.max_examples <= 0:
            return
    
        arguments = tuple(arguments)
    
        given_specifier = sd.tuples(
            sd.just(arguments),
            sd.fixed_dictionaries(generator_kwargs).map(
                lambda args: dict(args, **kwargs)
            )
        )
    
        def fail_health_check(message, label):
            if label in settings.suppress_health_check:
                return
            message += (
                '\nSee https://hypothesis.readthedocs.io/en/latest/health'
                'checks.html for more information about this. '
            )
            message += (
                'If you want to disable just this health check, add %s '
                'to the suppress_health_check settings for this test.'
            ) % (label,)
            raise FailedHealthCheck(message)
    
        search_strategy = given_specifier
        if selfy is not None:
            search_strategy = WithRunner(search_strategy, selfy)
    
        search_strategy.validate()
    
        perform_health_check = settings.perform_health_check
        perform_health_check &= Settings.default.perform_health_check
    
        from hypothesis.internal.conjecture.data import TestData, Status, \
            StopTest
        if not (
            Phase.reuse in settings.phases or
            Phase.generate in settings.phases
        ):
            return
    
        if perform_health_check:
            health_check_random = Random(random.getrandbits(128))
            # We "pre warm" the health check with one draw to give it some
            # time to calculate any cached data. This prevents the case
            # where the first draw of the health check takes ages because
            # of loading unicode data the first time.
            data = TestData(
                max_length=settings.buffer_size,
                draw_bytes=lambda data, n, distribution:
                distribution(health_check_random, n)
            )
            with Settings(settings, verbosity=Verbosity.quiet):
                try:
                    test_runner(data, reify_and_execute(
                        search_strategy,
                        lambda *args, **kwargs: None,
                    ))
                except BaseException:
                    pass
            count = 0
            overruns = 0
            filtered_draws = 0
            start = time.time()
            while (
                count < 10 and time.time() < start + 1 and
                filtered_draws < 50 and overruns < 20
            ):
                try:
                    data = TestData(
                        max_length=settings.buffer_size,
                        draw_bytes=lambda data, n, distribution:
                        distribution(health_check_random, n)
                    )
                    with Settings(settings, verbosity=Verbosity.quiet):
                        test_runner(data, reify_and_execute(
                            search_strategy,
                            lambda *args, **kwargs: None,
                        ))
                    count += 1
                except UnsatisfiedAssumption:
                    filtered_draws += 1
                except StopTest:
                    if data.status == Status.INVALID:
                        filtered_draws += 1
                    else:
                        assert data.status == Status.OVERRUN
                        overruns += 1
                except InvalidArgument:
                    raise
                except Exception:
                    if (
                        HealthCheck.exception_in_generation in
                        settings.suppress_health_check
                    ):
                        raise
                    report(traceback.format_exc())
                    if test_runner is default_new_style_executor:
                        fail_health_check(
                            'An exception occurred during data '
                            'generation in initial health check. '
                            'This indicates a bug in the strategy. '
                            'This could either be a Hypothesis bug or '
                            "an error in a function you've passed to "
                            'it to construct your data.',
                            HealthCheck.exception_in_generation,
                        )
                    else:
                        fail_health_check(
                            'An exception occurred during data '
                            'generation in initial health check. '
                            'This indicates a bug in the strategy. '
                            'This could either be a Hypothesis bug or '
                            'an error in a function you\'ve passed to '
                            'it to construct your data. Additionally, '
                            'you have a custom executor, which means '
                            'that this could be your executor failing '
                            'to handle a function which returns None. ',
                            HealthCheck.exception_in_generation,
                        )
            if overruns >= 20 or (
                not count and overruns > 0
            ):
                fail_health_check((
                    'Examples routinely exceeded the max allowable size. '
                    '(%d examples overran while generating %d valid ones)'
                    '. Generating examples this large will usually lead to'
                    ' bad results. You should try setting average_size or '
                    'max_size parameters on your collections and turning '
                    'max_leaves down on recursive() calls.') % (
                    overruns, count
                ), HealthCheck.data_too_large)
            if filtered_draws >= 50 or (
                not count and filtered_draws > 0
            ):
                fail_health_check((
                    'It looks like your strategy is filtering out a lot '
                    'of data. Health check found %d filtered examples but '
                    'only %d good ones. This will make your tests much '
                    'slower, and also will probably distort the data '
                    'generation quite a lot. You should adapt your '
                    'strategy to filter less. This can also be caused by '
                    'a low max_leaves parameter in recursive() calls') % (
                    filtered_draws, count
                ), HealthCheck.filter_too_much)
            runtime = time.time() - start
            if runtime > 1.0 or count < 10:
                fail_health_check((
                    'Data generation is extremely slow: Only produced '
                    '%d valid examples in %.2f seconds (%d invalid ones '
                    'and %d exceeded maximum size). Try decreasing '
                    "size of the data you're generating (with e.g."
                    'average_size or max_leaves parameters).'
                ) % (count, runtime, filtered_draws, overruns),
                    HealthCheck.too_slow,
                )
        last_exception = [None]
        repr_for_last_exception = [None]
    
        def evaluate_test_data(data):
            try:
                result = test_runner(data, reify_and_execute(
                    search_strategy, test,
                ))
                if result is not None and settings.perform_health_check:
                    fail_health_check((
                        'Tests run under @given should return None, but '
                        '%s returned %r instead.'
                    ) % (test.__name__, result), HealthCheck.return_value)
                return False
            except UnsatisfiedAssumption:
                data.mark_invalid()
            except (
                HypothesisDeprecationWarning, FailedHealthCheck,
                StopTest,
            ):
                raise
            except Exception:
                last_exception[0] = traceback.format_exc()
                verbose_report(last_exception[0])
                data.mark_interesting()
    
        from hypothesis.internal.conjecture.engine import TestRunner
    
        falsifying_example = None
        database_key = str_to_bytes(fully_qualified_name(test))
        start_time = time.time()
        runner = TestRunner(
            evaluate_test_data,
            settings=settings, random=random,
            database_key=database_key,
        )
        runner.run()
        run_time = time.time() - start_time
        timed_out = (
            settings.timeout > 0 and
            run_time >= settings.timeout
        )
        if runner.last_data is None:
            return
        if runner.last_data.status == Status.INTERESTING:
            falsifying_example = runner.last_data.buffer
            if settings.database is not None:
                settings.database.save(
                    database_key, falsifying_example
                )
        else:
            if runner.valid_examples < min(
                settings.min_satisfying_examples,
                settings.max_examples,
            ):
                if timed_out:
                    raise Timeout((
                        'Ran out of time before finding a satisfying '
                        'example for '
                        '%s. Only found %d examples in ' +
                        '%.2fs.'
                    ) % (
                        get_pretty_function_description(test),
>                       runner.valid_examples, run_time
                    ))
E                   Timeout: Ran out of time before finding a satisfying example for test_opp_win_attribution_rollup. Only found 2 examples in 91.21s.

.tox/py27-integration/local/lib/python2.7/site-packages/hypothesis/core.py:500: Timeout
```

**Comment by Daenyth** [comment]
Would you accept a PR using [this method](http://stackoverflow.com/questions/25188119/test-if-code-is-executed-from-within-a-py-test-session) to add a `hypothesis._called_from_pytest` boolean, and use that to `pytest.fail` for health checks instead of throwing an exception?

I'll also look at whether I can avoid the above entirely by using [__tracebackhide__ = True](http://doc.pytest.org/en/latest/example/simple.html#writing-well-integrated-assertion-helpers)

**Comment by alexwlchan** [comment]
Ugh, that is ugly. Thanks for the pull request!

I think we could probably do a better job of documenting the health checks as well, maybe by exposing the docstrings from https://github.com/HypothesisWorks/hypothesis-python/blob/master/src/hypothesis/errors.py, or writing some proper descriptions. I’ll look into it when I have a moment.

**Comment by Zac-HD** [comment]
> I'm planning on making some internal documentation for this as they come up but there's no reason it really has to be internal - how can I best submit things? It looks like you're not currently using the github wiki; is this something you'd want in a page for readthedocs?

Without having any idea what your internal documentation is like, the best place to add it is probably [docs/healthchecks.rst](https://github.com/HypothesisWorks/hypothesis-python/blob/master/docs/healthchecks.rst).  Is this still something you'd be willing to help out with?

**Comment by Zac-HD** [comment]
Closed in #973; both docs and traceback.
